src/openrct2/rct1/T4Importer.cpp

```diff
@@ -143,7 +143,7 @@ class TD4Importer final : public ITrackImporter
         td->ride_mode = td4Base.mode;
         if (td4Base.mode == RCT1_RIDE_MODE_POWERED_LAUNCH)
         {
-            td->ride_mode = RIDE_MODE_POWERED_LAUNCH;
+            td->ride_mode = static_cast<uint8_t>(RideMode::POWERED_LAUNCH);
```

Comment:
In this case I've had problems with `DataSerialiser` which couldn't take `enum class RideMode : uint8_t` as a parameter.
I'll change it and post more detailed info here.

Reply:
Ah okay i'll fix that when I get 12810 merged.

Reply:
Ok. I have made following changes:
file src/openrct2/ride/TrackDesign.h:
line 88 `uint8_t ride_mode;` to `RideMode ride_mode;`
src/openrct2/ride/TrackDesign.cpp:
line 112 `ride_mode = static_cast<uint8_t>(ride.mode);` to `ride_mode = ride.mode;`
src/openrct2/rct2/T6Importer.cpp:
line 80 `td->ride_mode = td6.ride_mode;` to `td->ride_mode = static_cast<RideMode>(td6.ride_mode);`
src/openrct2/rct1/T4Importer.cpp:
line 143 `td->ride_mode = td4Base.mode;` to `td->ride_mode = static_cast<RideMode>(td4Base.mode);`
line 146 `td->ride_mode = static_cast<uint8_t>(RideMode::POWERED_LAUNCH);` to `td->ride_mode = RideMode::POWERED_LAUNCH;`

and got following:
```
[  3%] Building CXX object CMakeFiles/libopenrct2.dir/src/openrct2/ride/coaster/InvertedHairpinCoaster.cpp.o
In file included from /home/lukasz/Programowanie/C++/OpenRCT2/src/openrct2/ride/../actions/../core/DataSerialiser.h:12,
                 from /home/lukasz/Programowanie/C++/OpenRCT2/src/openrct2/ride/../actions/GameAction.h:14,
                 from /home/lukasz/Programowanie/C++/OpenRCT2/src/openrct2/ride/../actions/FootpathPlaceFromTrackAction.hpp:23,
                 from /home/lukasz/Programowanie/C++/OpenRCT2/src/openrct2/ride/TrackDesign.cpp:16:
/home/lukasz/Programowanie/C++/OpenRCT2/src/openrct2/ride/../actions/../core/DataSerialiserTraits.h: In instantiation of ‘static void DataSerializerTraits<DataSerialiserTag<T> >::encode(OpenRCT2::IStream*, const DataSerialiserTag<T>&) [with T = RideMode]’:
/home/lukasz/Programowanie/C++/OpenRCT2/src/openrct2/ride/../actions/../core/DataSerialiser.h:82:67:   required from ‘DataSerialiser& DataSerialiser::operator<<(DataSerialiserTag<T>) [with T = RideMode]’
/home/lukasz/Programowanie/C++/OpenRCT2/src/openrct2/ride/TrackDesign.cpp:580:15:   required from here
/home/lukasz/Programowanie/C++/OpenRCT2/src/openrct2/ride/../actions/../core/DataSerialiserTraits.h:213:9: error: use of deleted function ‘static void DataSerializerTraits<T>::encode(OpenRCT2::IStream*, const T&) [with T = RideMode]’
  213 |         s.encode(stream, tag.Data());
      |         ^
/home/lukasz/Programowanie/C++/OpenRCT2/src/openrct2/ride/../actions/../core/DataSerialiserTraits.h:33:17: note: declared here
   33 |     static void encode(OpenRCT2::IStream* stream, const T& v) = delete;
      |                 ^~~~~~
/home/lukasz/Programowanie/C++/OpenRCT2/src/openrct2/ride/../actions/../core/DataSerialiserTraits.h: In instantiation of ‘static void DataSerializerTraits<DataSerialiserTag<T> >::decode(OpenRCT2::IStream*, DataSerialiserTag<T>&) [with T = RideMode]’:
/home/lukasz/Programowanie/C++/OpenRCT2/src/openrct2/ride/../actions/../core/DataSerialiser.h:84:67:   required from ‘DataSerialiser& DataSerialiser::operator<<(DataSerialiserTag<T>) [with T = RideMode]’
/home/lukasz/Programowanie/C++/OpenRCT2/src/openrct2/ride/TrackDesign.cpp:580:15:   required from here
/home/lukasz/Programowanie/C++/OpenRCT2/src/openrct2/ride/../actions/../core/DataSerialiserTraits.h:218:9: error: use of deleted function ‘static void DataSerializerTraits<T>::decode(OpenRCT2::IStream*, T&) [with T = RideMode]’
  218 |         s.decode(stream, tag.Data());
      |         ^
/home/lukasz/Programowanie/C++/OpenRCT2/src/openrct2/ride/../actions/../core/DataSerialiserTraits.h:34:17: note: declared here
   34 |     static void decode(OpenRCT2::IStream* stream, T& val) = delete;
      |                 ^~~~~~
/home/lukasz/Programowanie/C++/OpenRCT2/src/openrct2/ride/../actions/../core/DataSerialiserTraits.h: In instantiation of ‘static void DataSerializerTraits<DataSerialiserTag<T> >::log(OpenRCT2::IStream*, const DataSerialiserTag<T>&) [with T = RideMode]’:
/home/lukasz/Programowanie/C++/OpenRCT2/src/openrct2/ride/../actions/../core/DataSerialiser.h:88:60:   required from ‘DataSerialiser& DataSerialiser::operator<<(DataSerialiserTag<T>) [with T = RideMode]’
/home/lukasz/Programowanie/C++/OpenRCT2/src/openrct2/ride/TrackDesign.cpp:580:15:   required from here
/home/lukasz/Programowanie/C++/OpenRCT2/src/openrct2/ride/../actions/../core/DataSerialiserTraits.h:227:9: error: use of deleted function ‘static void DataSerializerTraits<T>::log(OpenRCT2::IStream*, const T&) [with T = RideMode]’
  227 |         s.log(stream, tag.Data());
      |         ^
/home/lukasz/Programowanie/C++/OpenRCT2/src/openrct2/ride/../actions/../core/DataSerialiserTraits.h:35:17: note: declared here
   35 |     static void log(OpenRCT2::IStream* stream, const T& val) = delete;
      |                 ^~~
```

Reply:
Can you please rebase your PR onto an updated develop and try doing this again, the PR that @duncanspumpkin mentioned was merged 3 days ago (1 day after you commented this), so I'm confident it will work now :D

Reply:
I've done a workaround for this in `src/openrct2/ride/TrackDesign.cpp` line 576:
`uint8_t ride_mode_uint=static_cast<uint8_t>(ride_mode);`
and in line 581:
`stream << DS_TAG(ride_mode_uint);`.

I have added also:
```
template<> struct DataSerializerTraits<RideMode> : public DataSerializerTraitsIntegral<RideMode>
{
};
```
in `/src/openrct2/core/DataSerialiserTraits.h` in line 112, but in order to use this I need guidance/explanation how this snippet works (same file, line 51 - 58):
```
    static void log(OpenRCT2::IStream* stream, const T& val)
    {
        std::stringstream ss;
        ss << std::hex << std::setw(sizeof(T) * 2) << std::setfill('0') << +val;

        std::string str = ss.str();
        stream->Write(str.c_str(), str.size());
    }
```
and if there is possibility that this will be changed in future?
In order to preserve this snippet there must be this workaround or the operators for every `enum class` must be overloaded.

Reply:
You shouldn't need to do that. With my PR now merged all enum classes should serialise correctly.

Reply:
My bad, I haven't done rebasing properly... <facepalm>